Operation t docs (#1902)
Azure/azure-sdk-for-cpp: doc/Operation.md

# Operation\<T>

> The Azure SDK for C++ defines an Operation as an abstract class to support a long running operation ([LRO](http://restalk-patterns.org/long-running-operation-polling.html)).

Reviewer: Rather than external (http) link, maybe we link to something like this instead:
https://github.com/microsoft/api-guidelines/blob/vNext/Guidelines.md#13-long-running-operations

> The Azure SDK for C++ defines an Operation as an abstract class to support a long running operation ([LRO](http://restalk-patterns.org/long-running-operation-polling.html)).

Reviewer: ```suggestion
The Azure SDK for C++ defines an `Operation<T>` templated abstract class to support a long running operation ([LRO](http://restalk-patterns.org/long-running-operation-polling.html)).
```

> The SDK clients would define a concrete class from the `Operation<T>` which an end users can consume for getting the status of the running operation.

Reviewer: Would the end-user ever defines these? This doc is for the end-user, I recommend talking directly to them as the audience.

```suggestion
Client libraries define specific concrete `Operation<T>` classes which you, as the end user calling the LRO API, would get back. This allows you to poll and get the status of your request.
```

Author: I don't think we have guidelines about the style for writing tech docs, do we?
I prefer not to use the `you ...` and at any case use something like `we ...` :)

## Operation Status

> The status of the operation defines the stages which the operation can be set from its creation and up to completion. The next table show the supported status of an `Operation<T>`.

Reviewer: ```suggestion
The status of the operation defines the stages which the operation can be in from its creation to completion. The following table shows the supported status of an `Operation<T>`.
```

| Status     | Description                                                                    |
| ---------- | ------------------------------------------------------------------------------ |
| NotStarted | The operation is created but the server has not yet started it.                |

> | Running    | The operation is running on the server side and has not yet be completed.      |

Reviewer: ```suggestion
| Running    | The operation is running on the server side and has not yet been completed.      |
```

| Succeeded  | The server has successfully finished running the LRO. This is a `final stage`. |
| Cancelled  | The operation was interrupted before completion. This is a `final stage`.      |
| Failed     | The server failed to run the LRO. This is a `final stage`.                     |

> > Note: A final stage means that the status won't change anymore.

Reviewer: Nice!

Author: @JeffreyRichter how would a service introduce more states if the OperationStatus is defined in Core?  Would they update the Core type for everybody?

Reviewer 2: The property is just a string. The service would return some string.

## Consuming an Operation\<T>

Any concrete class of `Operation<T>` will expose a `Poll()` and `PollUntilDone()` APIs. These methods will update the `operation status`. The status of an operation won't be automatically updated without calling `Poll()`.

### Poll

> The `Poll()` method will send a request to the Server to get the current status of the `LRO`. The Server will then return an Http raw response with information about the completion of the LRO. The Operation parses the response and updates its status. Finally, a read only reference to the http raw response (the one used to update the status) is returned by `Poll()`. We can use this raw response to parse specific information contained in the response such a progress, time, percentage or other data. The operation will only check for changes in the `operation status`, but the operation-consumer will know better about all the data coming back from the server in the response.

Author: I think it is important for the end-user to understand what `Poll` is expected to return and the reason it makes it like so

> The next code demonstrates the use of `Poll()`.

Author: back-tick adds a highlight format to the text and escapes html chars (like the `<`).
Using apostrophe would just display it as 'this is apostrophe',  but `this is back-tick format`.

```cpp
/**
 * Consuming Operation<T> interface to Poll for LRO updates.
 *
 */

> auto operationResult = client.ApiReturnOperation(...);

Author: My understanding is that not all operation would throw on init

> while (true)

Author: using `while(operation.IsDone())` is not convenience because calling `Poll` updates the m_status, se we need to check `IsDone()` after calling `Poll` which would duplicate the condition from the while and inside the while

{
    // Get a read-only ref to the Http raw response after polling. Poll will update the status.
    auto& response = operationResult.Poll();

    if(operationResult.IsDone()) {
        // Operation completed.
        break;
    }

    // Read the http headers in the response

>     for (auto& header : response.GetHeaders()) {

Author: What change do you have in mind?  My understanding is that we can't change the public surface any more, and that `Poll()` main reason to exists is for end-users to see the http raw-response of each `Poll`

        // ... consumer-code
    }
}

// Check is the Operation has created the value `T`
if (!operationResult.HasValue())
{
    // The Operation completed but it couldn't build the value `T`. Most likely the operation stage ended with error or it was cancelled.
    throw std::runtime_error("Operation finished with status:" + operationResult.Status());
}

// At this point, the Operation is success and we can get its value.
auto valueT = operationResult.Value();
```

### PollUntilDone

The `PollUntilDone()` is a convenience method for repeatedly call `Poll()` until the status is updated to a _final stage_. A `Response<T>` will be returned wrapping the model `T` and the Http raw response from which it was created. This method is useful when we are not interested about the Http raw response for each polling request.

The `PollUntilDone()` method requires an `std::chrono::milliseconds` input parameter which controls how much time to wait between calling `Poll`.

In the code below, instead of manually polling for updates, the `PollUntilDone` method is called.

```cpp
/**
 * Consuming Operation<T> interface. Calling PollUntilDone.
 *
 */

> auto operationResult = client.ApiReturnOperation(...);

Author: If we want it to compile, I would need to create an entire sample program (includes, concrete Operation class and main()). It would become too long IMO.

And yes, PollUntilDone will throw upon failing, but creating an Operation<T> depends on the concrete class if throws or not

try
{
    // Update each second
    auto response = operationResult.PollUntilDone(std::chrono::milliseconds(1000));

    // No need to check for `HasValue` here. The Operation will throw if the value can't be created after completing.

>     auto valueT = response.ExtractValue();

Reviewer 2: Were probably changing how this work a little bit.

Author: looking forward for it

    auto rawResponse = response.ExtractRawResponse();

}
catch (Azure::Core::RequestFailedException const& e)
{
    // Analyze the exception or re-throw.
}
```

## Operation\<T> lifetime

> The `Operation<T>` holds an Http raw response. Each Operation\<T> derived class defines the moment when the Http raw response is moved to the Operation. This means that an Operation can be created without initially holding an Http raw response. In this case, calling `GetRawResponse` method from the Operation will `throw` as no http raw response has been set for it yet.

Reviewer 2: We're probably getting rid of the raw response.

Author: We need to own an http raw response if we ever want to return a `Response<T>`.

The concrete Operation class defines since when the `GetRawResponse` can be called, we need to review each derived class to know at what time it is valid or not to call `GetRawResponse`.

The code below show two examples of Operation derive classes. The first one requires an http raw response to be created, ensuring that `GetRawResponse` can be called at any moment after the Operation is created. The second class does not set an http raw response when it is init, which means that `GetRawResponse` can only be called after calling `Poll`, otherwise it will throw.

```cpp
/**
 * Example of operations construction and `GetRawResponse` valid calls.
 *
 */

// The operation can only be constructed based on an http raw response.
class OperationA : public Azure::Core::Operation<Model> {
public:

>     OperationA(Azure::Core::Http::RawResponse rawResponse);

Reviewer 2: This doesn't seem right to me. Only the client lib should be able to construct an Operation<T> and it will most-likely NOT be created from a raw response. Customers don't need to know about this - move to guidelines.

Author: I was thinking that the `Operation<T>` gives a powerful feature and advanced feature to the users. They can basically create `Future requests` with it :) if they want.  But that's fine, I just wanted to make the general knowledge of this Class to have official documentation to support it.  I can move this else where
